onnxruntime/core/providers/cuda/math/topk.cc

```diff
@@ -77,19 +77,28 @@ Status TopK<inputk>::ComputeInternal(OpKernelContext* ctx) const {
   int32_t axis = static_cast<int32_t>(axis_ < 0 ? rank + axis_ : axis_);
   ORT_ENFORCE(axis > -1 && axis < rank);
 
+  int64_t k_value = 0;
   if (inputk) {
     auto tensor_K = ctx->Input<Tensor>(1);
     ORT_ENFORCE(nullptr != tensor_K);
-    K_ = *tensor_K->Data<int64_t>();
-    ORT_ENFORCE(K_ >= 0 && K_ <= tensor_X->Shape().GetDims()[axis]);
+    k_value = *tensor_K->Data<int64_t>();
+  } else {  // from attribute
+    k_value = attr_k_;
+  }
+
+  // Now that we know the value of 'K' and the input shape,
+  // make a final validation before going to the implementation
+  if ((k_value < 0) || (k_value > tensor_X->Shape().GetDims()[axis])) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Value of K outside range. K value: ", k_value,
+                           ". Input shape: ", tensor_X->Shape(), " . Axis: ", axis);
   }
 
   auto output_shape = tensor_X->Shape();
-  output_shape[axis] = K_;
+  output_shape[axis] = k_value;
   auto tensor_V = ctx->Output(0, output_shape);
   auto tensor_I = ctx->Output(1, output_shape);
 
-  if (0 == K_) {
+  if (output_shape.Size() == 0) { // Bail out early if the output is going to be empty 
```

Comment:
This issue was investigated and fixed by @er3x3 in https://github.com/microsoft/onnxruntime/pull/19914 - included that fix here because I had to make changes to this line anyway.